pulsar RACE issue

I'm from Milvus team, we use Pulsar in our system.
When I run unittest, I often see following pulsar RACE issue.
Can you help to fix this?

```
WARNING: DATA RACE
Read at 0x00c00015de68 by goroutine 144:
  github.com/apache/pulsar-client-go/pulsar.(*partitionConsumer).internalClose()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_partition.go:842 +0x5f1
  github.com/apache/pulsar-client-go/pulsar.(*partitionConsumer).runEventsLoop()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_partition.go:801 +0x29b

Previous write at 0x00c00015de68 by goroutine 208:
  github.com/apache/pulsar-client-go/pulsar.(*partitionConsumer).Decompress()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_partition.go:1062 +0x468
  github.com/apache/pulsar-client-go/pulsar.(*partitionConsumer).MessageReceived()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_partition.go:474 +0x104
  github.com/apache/pulsar-client-go/pulsar/internal.(*connection).handleMessage()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/internal/connection.go:638 +0x1ab
  github.com/apache/pulsar-client-go/pulsar/internal.(*connection).internalReceivedCommand()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/internal/connection.go:527 +0x472
  github.com/apache/pulsar-client-go/pulsar/internal.(*connection).run()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/internal/connection.go:381 +0x57c
  github.com/apache/pulsar-client-go/pulsar/internal.(*connection).start.func1()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/internal/connection.go:215 +0xc4

Goroutine 144 (running) created at:
  github.com/apache/pulsar-client-go/pulsar.newPartitionConsumer()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_partition.go:207 +0xf35
  github.com/apache/pulsar-client-go/pulsar.(*consumer).internalTopicSubscribeToPartitions.func1()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_impl.go:315 +0x778

Goroutine 208 (running) created at:
  github.com/apache/pulsar-client-go/pulsar/internal.(*connection).start()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/internal/connection.go:211 +0x4c
  github.com/apache/pulsar-client-go/pulsar/internal.(*connectionPool).GetConnection()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/internal/connection_pool.go:100 +0x6de
  github.com/apache/pulsar-client-go/pulsar/internal.(*rpcClient).Request()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/internal/rpc_client.go:112 +0xed
  github.com/apache/pulsar-client-go/pulsar.(*partitionConsumer).grabConn()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_partition.go:951 +0x15d1
  github.com/apache/pulsar-client-go/pulsar.newPartitionConsumer()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_partition.go:179 +0xc51
  github.com/apache/pulsar-client-go/pulsar.(*consumer).internalTopicSubscribeToPartitions.func1()
      /go/pkg/mod/github.com/apache/pulsar-client-go@v0.4.0/pulsar/consumer_impl.go:315 +0x778
==================
time="2021-05-09T16:12:45Z" level=info msg="[Connection closed]" local_addr="172.18.0.5:39286" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Error reading from connection]" error="Short read when reading frame size: read tcp 172.18.0.5:39282->172.18.0.4:6650: use of closed network connection" local_addr="172.18.0.5:39282" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Connection closed]" local_addr="172.18.0.5:39292" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Error reading from connection]" error="Short read when reading frame size: read tcp 172.18.0.5:39286->172.18.0.4:6650: use of closed network connection" local_addr="172.18.0.5:39286" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Error reading from connection]" error="Short read when reading frame size: read tcp 172.18.0.5:39292->172.18.0.4:6650: use of closed network connection" local_addr="172.18.0.5:39292" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Connection closed]" local_addr="172.18.0.5:39284" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Connection closed]" local_addr="172.18.0.5:39288" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Error reading from connection]" error="Short read when reading frame size: read tcp 172.18.0.5:39284->172.18.0.4:6650: use of closed network connection" local_addr="172.18.0.5:39284" remote_addr="pulsar://pulsar:6650"
time="2021-05-09T16:12:45Z" level=info msg="[Error reading from connection]" error="Short read when reading frame size: read tcp 172.18.0.5:39288->172.18.0.4:6650: use of closed network connection" local_addr="172.18.0.5:39288" remote_addr="pulsar://pulsar:6650"
--- FAIL: TestAssignSegmentID (0.40s)
    testing.go:1042: race detected during execution of test
FAIL
coverage: 68.0% of statements
FAIL	github.com/milvus-io/milvus/internal/dataservice	4.807s
FAIL
make: *** [test-go] Error 1
Makefile:149: recipe for target 'test-go' failed
2
Error: Process completed with exit code 2.
```

--- a/pulsar/consumer_partition.go
+++ b/pulsar/consumer_partition.go
@@ -138,6 +138,7 @@ type partitionConsumer struct {
 
 	log log.Logger
 
+	providersMutex       sync.Mutex
 	compressionProviders map[pb.CompressionType]compression.Provider
 	metrics              *internal.TopicMetrics
 }
@@ -850,9 +851,11 @@ func (pc *partitionConsumer) internalClose(req *closeRequest) {
 		pc.log.Info("Closed consumer")
 	}
 
+	pc.providersMutex.Lock()
 	for _, provider := range pc.compressionProviders {
 		provider.Close()
 	}
+	pc.providersMutex.UnLock()
 
 	pc.setConsumerState(consumerClosed)
 	pc.conn.DeleteConsumeHandler(pc.consumerID)
@@ -1062,6 +1065,7 @@ func getPreviousMessage(mid trackingMessageID) trackingMessageID {
 }
 
 func (pc *partitionConsumer) Decompress(msgMeta *pb.MessageMetadata, payload internal.Buffer) (internal.Buffer, error) {
+	pc.providersMutex.Lock()
 	provider, ok := pc.compressionProviders[msgMeta.GetCompression()]
 	if !ok {
 		var err error
@@ -1072,6 +1076,7 @@ func (pc *partitionConsumer) Decompress(msgMeta *pb.MessageMetadata, payload int
 
 		pc.compressionProviders[msgMeta.GetCompression()] = provider
 	}
+	pc.providersMutex.UnLock()
 
 	uncompressed, err := provider.Decompress(nil, payload.ReadableSlice(), int(msgMeta.GetUncompressedSize()))
 	if err != nil {

--- a/pulsar/consumer_partition.go
+++ b/pulsar/consumer_partition.go
@@ -138,7 +138,7 @@ type partitionConsumer struct {
 
 	log log.Logger
 
-	providersMutex       sync.Mutex
+	providersMutex       sync.RWMutex
 	compressionProviders map[pb.CompressionType]compression.Provider
 	metrics              *internal.TopicMetrics
 }
@@ -855,7 +855,7 @@ func (pc *partitionConsumer) internalClose(req *closeRequest) {
 	for _, provider := range pc.compressionProviders {
 		provider.Close()
 	}
-	pc.providersMutex.UnLock()
+	pc.providersMutex.Unlock()
 
 	pc.setConsumerState(consumerClosed)
 	pc.conn.DeleteConsumeHandler(pc.consumerID)
@@ -1065,19 +1065,20 @@ func getPreviousMessage(mid trackingMessageID) trackingMessageID {
 }
 
 func (pc *partitionConsumer) Decompress(msgMeta *pb.MessageMetadata, payload internal.Buffer) (internal.Buffer, error) {
-	pc.providersMutex.Lock()
+	pc.providersMutex.RLock()
 	provider, ok := pc.compressionProviders[msgMeta.GetCompression()]
+	pc.providersMutex.RUnlock()
 	if !ok {
 		var err error
 		if provider, err = pc.initializeCompressionProvider(msgMeta.GetCompression()); err != nil {
 			pc.log.WithError(err).Error("Failed to decompress message.")
-			pc.providersMutex.UnLock()
 			return nil, err
 		}
 
+		pc.providersMutex.Lock()
 		pc.compressionProviders[msgMeta.GetCompression()] = provider
+		pc.providersMutex.Unlock()
 	}
-	pc.providersMutex.UnLock()
 
 	uncompressed, err := provider.Decompress(nil, payload.ReadableSlice(), int(msgMeta.GetUncompressedSize()))
 	if err != nil {

--- a/pulsar/consumer_partition.go
+++ b/pulsar/consumer_partition.go
@@ -1071,6 +1071,7 @@ func (pc *partitionConsumer) Decompress(msgMeta *pb.MessageMetadata, payload int
 		var err error
 		if provider, err = pc.initializeCompressionProvider(msgMeta.GetCompression()); err != nil {
 			pc.log.WithError(err).Error("Failed to decompress message.")
+			pc.providersMutex.UnLock()
 			return nil, err
 		}
 

